sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicOperators.scala

```diff
@@ -180,6 +229,55 @@ case class Join(
     }
   }
 
+  override protected def validConstraints: Set[Expression] = {
+    // Currently we only propagate constraints if the condition consists of equality
+    // and ranges. For all other cases, we return an empty set of constraints
+    def constructIsNotNullConstraints(condition: Expression): Set[Expression] = {
+      splitConjunctivePredicates(condition).map {
+        case EqualTo(l, r) =>
+          Set(IsNotNull(l), IsNotNull(r))
+        case GreaterThan(l, r) =>
+          Set(IsNotNull(l), IsNotNull(r))
+        case GreaterThanOrEqual(l, r) =>
+          Set(IsNotNull(l), IsNotNull(r))
+        case LessThan(l, r) =>
+          Set(IsNotNull(l), IsNotNull(r))
+        case LessThanOrEqual(l, r) =>
+          Set(IsNotNull(l), IsNotNull(r))
+        case _ =>
+          Set.empty[Expression]
+      }.foldLeft(Set.empty[Expression])(_ union _.toSet)
+    }
+
+    def constructIsNullConstraints(plan: LogicalPlan): Set[Expression] = {
+      plan.output.map(IsNull).toSet
+    }
+
+    (joinType match {
+      case Inner if condition.isDefined =>
+        getRelevantConstraints(left)
+          .union(getRelevantConstraints(right))
+          .union(constructIsNotNullConstraints(condition.get))
+      case LeftSemi if condition.isDefined =>
+        getRelevantConstraints(left)
+          .union(getRelevantConstraints(right))
+          .union(constructIsNotNullConstraints(condition.get))
+      case LeftOuter =>
+        getRelevantConstraints(left)
+          .union(constructIsNullConstraints(right))
+      case RightOuter =>
+        getRelevantConstraints(right)
+          .union(constructIsNullConstraints(left))
```

Comment:
Maybe I'm not following what this does, but we don't know that the columns from the left _are_ `null`, they just might be `null`.